src/notifyme.js

```diff
@@ -1,116 +1,117 @@
+     *      placed here to simplify the process for users who just want to use
+     *      the default options.
+     *  @param {Object} `options`   The options parameter for the notification
+     *      instance and is used to extend the default options.
+     */
+    function NotifyMe(title, body, options) {
+        if (!(this instanceof NotifyMe)) {
+            return new NotifyMe(title, body, options);
+        }
+        this.title = title;
+        this.options = extend(defaultOptions, options || {});
+        this.options.body = body;
+        this.options.tag = generateTag();
+    }
 
-        /**
-         * Checks for the permission from the user
-         * @param  {String} title   Title for Notification
-         * @param  {Object} options Object with Options for Notification
-         * @return {NA}         NA
-         */
-        getPermissionAndLaunch = function (title, options) {
-            Notification.requestPermission(function (permission) {
-                if (permission === 'granted') {
-                    launchNotiWindow(title, mergeObjects(defaultOptions, options));
-                } else if (permission === 'denied') {
-                    console.log("Request denied by the User");
+    NotifyMe.permissionGranted = false;
+
+    /**
+     *  Request permission from the user to use the Notifications API.
+     *  @return {Promise}
+     */
+    NotifyMe.requestPermission = function() {
+        return new Promise(function(resolve, reject) {
+            Notification.requestPermission(function(permission) {
+                NotifyMe.permissionGranted = permission === 'granted';
+                if (NotifyMe.permissionGranted) {
+                    resolve();
+                } else {
+                    reject();
                 }
             });
-        };
-        /**
-         * Prototype function for invoking the Notification object.
-         * @return {NA} [NA]
-         */
-        Notify.prototype.launch = function () {
-            return getPermissionAndLaunch(this.title, this.options);
-        };
-        return Notify;
-    })();
+        });
+    };
 
-    NotifyMe.launch = function (title, options) {
-        var tempObj = new Notify(title, options);
-        tempObj.launch();
+    /**
+     *  Launch a notification instance, asking permission from the user first
+     *  when permission is not yet granted.
+     *  @return {Promise}
+     */
+    NotifyMe.prototype.launch = function() {
+        if (NotifyMe.permissionGranted) {
+            this.notification = new Notification(this.title, this.options);
+            this.notification.onclick = this.options.onclick;
+            this.notification.onerror = this.options.onerror;
+            this.notification.onclose = this.options.onclose;
+            return Promise.resolve(this.notification);
+        } else {
+            return NotifyMe.requestPermission()
+                .then(this.launch.bind(this))
```

Comment:
I need to `.bind(this)` so that the context of `this` inside the second call to `launch` will still be the `NotifyMe` instance. If `.bind(this)` is removed the `this` inside the second call becomes the `window` object.

Btw thanks for all your comments, I'll be updating my pull request to incorporate them :+1: